client/src/partials/templates/modals/findReference.modal.js

```diff
@@ -0,0 +1,151 @@
+  'CreditorService', 'VoucherService', 'JournalFilteringService', 'data'
+];
+
+/**
+ * Find Reference Modal Controller
+ *
+ * This controller provides bindings for the find references modal.
+ * @todo Implement the Patient Invoices Data list for the references
+ * @todo Implement the Cash Payment Data list for the references
+ * @todo Implement the link to the document for exactitude of the reference
+ */
+function FindReferenceModalController($scope, Instance, Debtor, Creditor, Voucher, Filtering, Data) {
+  var vm = this;
+
+  vm.result = {};
+
+  vm.documentType = {
+    'patient_invoice' : {
+      label : 'VOUCHERS.COMPLEX.PATIENT_INVOICE',
+      action : referencePatientInvoice
+    },
+    'cash_payment' : {
+      label : 'VOUCHERS.COMPLEX.CASH_PAYMENT',
+      action : referenceCashPayment
+    },
+    'voucher' : {
+      label : 'VOUCHERS.COMPLEX.VOUCHER',
+      action : referenceVoucher
+    }
+  }
+
+  vm.selectDocType = selectDocType;
+  vm.submit  = submit;
+  vm.cancel  = cancel;
+  vm.refresh = refresh;
+
+  /** Grid configurations */
+  vm.gridOptions = {};
+
+  /** grid utilities */
+  var filtering  = new Filtering(vm.gridOptions);
+
+  vm.gridOptions.multiSelect     = false;
+  vm.gridOptions.enableFiltering = true;
+  vm.gridOptions.onRegisterApi   = onRegisterApi;
+
+  function onRegisterApi(gridApi) {
+    vm.gridApi = gridApi;
+    /** @fixme: use of $scope is not optimal */
+    vm.gridApi.selection.on.rowSelectionChanged($scope, rowSelectionCallback);
+  }
+
+  function rowSelectionCallback(row) {
+    vm.selectedRow = row.entity;
+  }
+  /** End grid configuration */
+
+  // startup the modal
+  startup();
+
+  Debtor.read()
+  .then(function (list) {
+    vm.debtorList = list;
+  });
+
+  Creditor.read()
+  .then(function (list) {
+    vm.creditorList = list;
+  });
+
+  function referencePatientInvoice() {
+    /**
+    * @fixme the balance value is not correct for the document references
+    * The correct value is the amount of the invoice in the receipt
+    */
+    return;
+  }
+
+  function referenceCashPayment() {
+    /**
+    * @fixme the balance value is not correct for the document references
+    * The correct value is the amount of the invoice in the receipt
+    */
+    return;
+  }
+
+  function referenceVoucher() {
+    Voucher.read()
+    .then(function (list) {
+      vm.gridOptions.columnDefs  = [
+        { field : 'reference', displayName : 'Reference'},
+        { field : 'description', displayName : 'Description'},
+        { field : 'amount', displayName : 'Amount'},
+        {
+          field : 'date',
+          displayName : 'Date',
+          cellFilter : 'date:"mediumDate"',
+          filter : { condition : filtering.byDate }
+        },
+      ];
+      /**
+       * @fixme: the voucher API returns two lines
+       * for credit and debit with the same amount; and we need
+       * only one line with the amount
+       * we choose to discare the credit line
```

Comment:
This is an excellent comment.  Choosing the credit line is perfectly acceptable.